File: leetcode/learn-graphs-DFS__Find-if-Path-Exists-in-Graph.js
/**
 * @param {number} n
 * @param {number[][]} edges
 * @param {number} source
 * @param {number} destination
 * @return {boolean}
 */
var validPath = function (n, edges, source, destination) {
    const graph = new Map()
    const visited = []

    for (let i = 0; i < n; i++) {
        graph.set(i, [])
        visited.push(false)
    }

    for (let i = 0; i < edges.length; i++) {
        const [from, to] = edges[i]

        graph.set(from, [...graph.get(from), to])
        graph.set(to, [...graph.get(to), from])

        vs

        graph.get(from).push(to)
        graph.get(to).push(from)
    }

    // console.log(graph)

    const stack = [destination]

    visited[destination] = true

    while (stack.length) {
        const node = stack.pop()

        if (node === source) {
            return true
        }

        const neighbours = graph.get(node)

        for (let i = 0; i < neighbours.length; i++) {
            const neighbour = neighbours[i]

            if (!visited[neighbour]) {
                if (neighbour === source) {
                    return true
                }

                stack.push(neighbour)
                visited[neighbour] = true
            }
        }
    }

    return false
};

{
    const n = 3
    const edges = [
        [0, 1],
        [1, 2],
        [2, 0]
    ]
    const source = 0
    const destination = 2
    const expected = true
    const result = validPath(n, edges, source, destination)

    console.log(expected === result)
    console.log('------------------')
}

{
    const n = 6
    const edges = [
        [0, 1],
        [0, 2],
        [3, 5],
        [5, 4],
        [4, 3]
    ]
    const source = 0
    const destination = 5
    const expected = false
    const result = validPath(n, edges, source, destination)

    console.log(expected === result)
    console.log('------------------')
}

File: leetcode/learn-graphs-DFS-Find-if-Path-Exists-in-Graph.py
from typing import List
from collections import deque


# @lc code=start
class Solution:
    def validPath(self, n: int, edges: List[List[int]], source: int, destination: int) -> bool:
        graph = [[] for _ in range(n)]

        for u,v in edges:
            graph[u].append(v)
            graph[v].append(u)

        print(graph)

        stack = []
        visited = [False for _ in range(n)]

        stack.append(source)

        while len(stack) > 0:
            node = stack.pop()

            if node == destination:
                return True
            
            if visited[node]:
                continue

            visited[node] = True

            for neighbour in graph[node]:
                stack.append(neighbour)

        return False
# @lc code=end

'''
'''


testCases = [
    {
        "n": 3,
        "edges": [[0,1],[1,2],[2,0]],
        "source": 0,
        "destination": 2,
        "expected": True
    },
    {
        "n": 6,
        "edges": [[0,1],[0,2],[3,5],[5,4],[4,3]],
        "source": 0,
        "destination": 5,
        "expected": False
    },
]

for testCase in testCases:
    print('')

    n = testCase["n"]
    edges = testCase["edges"]
    source = testCase["source"]
    destination = testCase["destination"]
    expected = testCase["expected"]

    s = Solution()

    result = s.validPath(n, edges, source, destination)
    print(n, edges, source, destination)
    assert result == expected, f"result {result} should be expected: {expected}"
